tests/components/bmw_connected_drive/test_diagnostics.py

```diff
@@ -0,0 +1,98 @@
+"""Test BMW diagnostics."""
+
+import datetime
+import json
+import os
+import time
+
+from freezegun import freeze_time
+import pytest
+
+from homeassistant.components.bmw_connected_drive.const import DOMAIN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr
+
+from . import setup_mocked_integration
+
+from tests.common import load_fixture
+from tests.components.diagnostics import (
+    get_diagnostics_for_config_entry,
+    get_diagnostics_for_device,
+)
+
+
+@freeze_time(datetime.datetime(2022, 7, 10, 11))
+async def test_config_entry_diagnostics(hass: HomeAssistant, hass_client, bmw_fixture):
+    """Test config entry diagnostics."""
+
+    # Make sure that local timezone for test is UTC
+    os.environ["TZ"] = "UTC"
+    time.tzset()
+
+    mock_config_entry = await setup_mocked_integration(hass)
+
+    diagnostics = await get_diagnostics_for_config_entry(
+        hass, hass_client, mock_config_entry
+    )
+
+    diagnostics_fixture = json.loads(
+        load_fixture("diagnostics/diagnostics_config_entry.json", DOMAIN)
+    )
+
+    assert diagnostics == diagnostics_fixture
+
+
+@freeze_time(datetime.datetime(2022, 7, 10, 11))
+async def test_device_diagnostics(hass: HomeAssistant, hass_client, bmw_fixture):
+    """Test device diagnostics."""
+
+    # Make sure that local timezone for test is UTC
+    os.environ["TZ"] = "UTC"
+    time.tzset()
+
+    mock_config_entry = await setup_mocked_integration(hass)
+
+    device_registry = dr.async_get(hass)
+    reg_device = device_registry.async_get_device(
+        identifiers={(DOMAIN, "WBY00000000REXI01")},
+    )
+    assert reg_device is not None
+
+    diagnostics = await get_diagnostics_for_device(
+        hass, hass_client, mock_config_entry, reg_device
+    )
+
+    diagnostics_fixture = json.loads(
+        load_fixture("diagnostics/diagnostics_device.json", DOMAIN)
+    )
+
+    assert diagnostics == diagnostics_fixture
+
+
+@freeze_time(datetime.datetime(2022, 7, 10, 11))
+async def test_device_diagnostics_vehicle_not_found(
+    hass: HomeAssistant, hass_client, bmw_fixture
+):
+    """Test device diagnostics when the vehicle cannot be found."""
+
+    # Make sure that local timezone for test is UTC
+    os.environ["TZ"] = "UTC"
+    time.tzset()
+
+    mock_config_entry = await setup_mocked_integration(hass)
+
+    device_registry = dr.async_get(hass)
+    reg_device = device_registry.async_get_device(
+        identifiers={(DOMAIN, "WBY00000000REXI01")},
+    )
+    assert reg_device is not None
+
+    # Change vehicle identifier so that vehicle will not be found
+    device_registry.async_update_device(
+        reg_device.id, new_identifiers={(DOMAIN, "WBY00000000REXI99")}
+    )
+
+    with pytest.raises(AssertionError):
```

Comment:
Where is `AssertionError` raised?

Reply:
It is raised by the diagnostics helper function if a device is does not exist (HTTPStatus != OK due to the `HomeAssistantError` raised if a device is not found by our diagnostics).

```
    async def _get_diagnostics_for_device(hass, hass_client, config_entry, device):
        """Return the diagnostics for the specified device."""
        assert await async_setup_component(hass, "diagnostics", {})
    
        client = await hass_client()
        response = await client.get(
            f"/api/diagnostics/config_entry/{config_entry.entry_id}/device/{device.id}"
        )
>       assert response.status == HTTPStatus.OK
E       AssertionError

tests/components/diagnostics/__init__.py:33: AssertionError
```

Reply:
I'm checking if we should go ahead with this for now or if we want to change something in diagnostics integration to handle this better first.

Reply:
To be honest, I only included it because the [subaru integration](https://github.com/home-assistant/core/blob/dev/homeassistant/components/subaru/diagnostics.py#L39-L56) is built the same.

However we also could remove it completely - the only reasons why a vehicle as loaded device cannot be found (i.e. why the `HomeAssistantError` could be raised) are:
- BMW API is unavailable
- car was removed from the MyBMW account

In case `get_vehicles()` tries to get a vehicle that is not available anymore, it will return `None` and everything should work without any issues (only `diagnostics_data.data` will be `None` which is correct).

Reply:
Tested not raising `HomeAssistantError`. Works as expected and therefore removed the exception handling.